### DISCUSSION TITLE
Is it possible to create DB and then pass it's connection string as a build secret to another resource built from Dockerfile?

### DISCUSSION THREAD

**Discussion by megafinz** [microsoft/aspire#6897] [Q&A] [answered]
Hey there.

I have a NextJS website that requires access to a database during the build phase (static site generation).

Is it possible to arrange such setup using Aspire? Basically I need to create a database and somehow pass it's connection string as a build secret to docker build. Seems like `WithBuildSecret` accepts only `ParameterResource` and I'm having trouble figuring out how can marry two things together (if this is possible at all of course).

Any help appreciated.

EDIT:

I suppose as a workaround I can create username and password myself and use them to create the database resource and then construct the connection string manually, but still curious if it is possible to use opaque connection string generated by the database resource itself.

EDIT 2:

Ugh, figured out how to manually construct the connection string (using `ReferenceExpression`), but can't figure out how to transform that into `ParameterResource` :(

**Comment by davidfowl** [comment]
> I suppose as a workaround I can create username and password myself and use them to create the database resource and then construct the connection string manually, but still curious if it is possible to use opaque connection string generated by the database resource itself.

Looking at the code, it should be possible if you use the lower-level annotation to make this work:

```C#
public static IResourceBuilder<T> WithBuildSecret<T>(this IResourceBuilder<T> builder, string name, IResourceBuilder<IResourceWithConnectionString> value) where T : ContainerResource
{
    ArgumentNullException.ThrowIfNull(builder);
    ArgumentException.ThrowIfNullOrEmpty(name);
    ArgumentNullException.ThrowIfNull(value);

    var annotation = builder.Resource.Annotations.OfType<DockerfileBuildAnnotation>().SingleOrDefault();

    if (annotation is null)
    {
        throw new InvalidOperationException("The resource does not have a Dockerfile build annotation. Call WithDockerfile before calling WithSecretBuildArg.");
    }

    annotation.BuildSecrets[name] = value.Resource;

    return builder;
}
```

PS: That said, I'm not sure if this is the "right" thing to do.

**Comment by megafinz** [comment]
To avoid complicating the code decided to solve this with some config duplication and just provide username/password and connection strings as separate `ParameterResource`s.

But now there's another issue: it seems that either network resources are inaccessible inside container build process or container build process doesn't respect DB healthchecks.

E.g. in this example `RUN pg_isready -U postgres -h db` will fail during the container build phase. If I remove this line and then execute it inside running container then it will succeeed.

```csharp
// Program.cs
var builder = DistributedApplication.CreateBuilder(args);
var db = builder.AddPostgres("db");
var pgTest = builder.AddDockerfile("pg-test", "./").WaitFor(db);
builder.Build().Run();
```

```dockerfile
# Dockerfile
FROM httpd
RUN apt update
RUN apt install -y postgresql
RUN pg_isready -U postgres -h db

```

**Reply by megafinz** [reply, marked as answer]
> But now there's another issue: it seems that either network resources are inaccessible inside container build process or container build process doesn't respect DB healthchecks.

Mkay, after some more investigation it turns out that my build-time connectivity issues likely stem from [the way Docker networking works during build](https://docs.docker.com/reference/cli/docker/buildx/build/#network). 

TLDR: you can't really access other resources during build, even when using docker-compose you need to resolve to some [hacks involving the usage of host's network](https://github.com/docker/compose/issues/6148#issuecomment-1273690631).
